### PR TITLE
Feature/#29 detect enemies within tower range

### PR DESCRIPTION
**Pull request by DoctorDarius**
Also includes #30 Towers shoots at enemies automatically

### CODE REVIEW

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+namespace Assets
+{
+    public class TowerAI : MonoBehaviour
+    {
+        public Rigidbody Projectile;
+        public Transform ProjectileSpawn;
+        public Transform Target = null;
+        private IEnumerator _coroutine;
+        public float ReloadTime;
+        public float BulletSpeed;
```
Comment (hassanAlAwwadi) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: Please add default values for bulletspeed and releadtime

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
+            if (enemyScript == null) return;
+            Target = target.transform;
+        }
+
+        void OnTriggerExit(Collider target)
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
+            if (enemyScript == null) return;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: Can we add newlines around if statements please.

```suggestion

            if (enemyScript == null) return;

```

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        public Transform ProjectileSpawn;
+        public Transform Target = null;
+        private IEnumerator _coroutine;
+        public float ReloadTime = 2;
+        public float BulletSpeed = 5;
+
+        void OnTriggerEnter(Collider target)
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
+            if (enemyScript == null) return;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]:
```suggestion

            if (enemyScript == null) return;

```

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            Target = null;
+        }
+
+
+        // Use this for initialization
+        void Start()
+        {
+            _coroutine = Reload(ReloadTime);
+            StartCoroutine(_coroutine);
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: I think we need to stop the coroutine when the gameobject instance is destoyed 🤔.

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            while (true)
+            {
+                AimAndShoot();
+                yield return new WaitForSeconds(waitTime);
+            }
+        }
+
+        void AimAndShoot()
+        {
+            if (Target == null) return;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]:
```suggestion
            if (Target == null) return;

```

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+        void OnTriggerEnter(Collider target)
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
+            if (enemyScript == null) return;
+            Target = target.transform;
+        }
+
+        void OnTriggerExit(Collider target)
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]:
```suggestion
            var enemyScript = target.gameObject.GetComponent<Enemy>();
```

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+        public Rigidbody Projectile;
+        public Transform ProjectileSpawn;
+        public Transform Target = null;
+        private IEnumerator _coroutine;
+        public float ReloadTime = 2;
+        public float BulletSpeed = 5;
+
+        void OnTriggerEnter(Collider target)
+        {
+            Enemy enemyScript = target.gameObject.GetComponent<Enemy>();
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]:
```suggestion
            var enemyScript = target.gameObject.GetComponent<Enemy>();
```

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+using UnityEngine;
+
+namespace Assets
+{
+    public class TowerAI : MonoBehaviour
+    {
+        public Rigidbody Projectile;
+        public Transform ProjectileSpawn;
+        public Transform Target = null;
+        private IEnumerator _coroutine;
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: Can we separate private and public properties please 🙂.

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+
+        void AimAndShoot()
+        {
+            if (Target == null) return;
+            var newProjectile = (Rigidbody)Instantiate(Projectile, ProjectileSpawn.position, Projectile.rotation);
+            newProjectile.velocity = (Target.transform.position - transform.position).normalized * BulletSpeed;
+        }
+
+        // Update is called once per frame
+        void Update()
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: This empty method can be removed.

File: src/Assets/LightSkyDefense/Scripts/TowerAI.cs [hunk 10]
```diff
@@ -0,0 +1,58 @@
+﻿using System.Collections;
+using UnityEngine;
+
+namespace Assets
+{
+    public class TowerAI : MonoBehaviour
```
Comment (leroykorterink) [inline src/Assets/LightSkyDefense/Scripts/TowerAI.cs]: Can we rename `TowerAI` to `TowerBehaviour`.

**Pull request merged**